staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -804,11 +804,42 @@ func (qs *queueSet) finishRequestAndDispatchAsMuchAsPossible(req *request) bool
 func (qs *queueSet) finishRequestLocked(r *request) {
 	now := qs.clock.Now()
 	qs.totRequestsExecuting--
-	qs.totSeatsInUse -= r.Seats()
 	metrics.AddRequestsExecuting(r.ctx, qs.qCfg.Name, r.fsName, -1)
-	metrics.AddRequestConcurrencyInUse(qs.qCfg.Name, r.fsName, -r.Seats())
 	qs.obsPair.RequestsExecuting.Add(-1)
 
+	// TODO: for now we keep the logic localized so it is easier to see
+	//  how the counters are tracked for queueset and queue, in future we
+	//  can refactor to move this function.
+	releaseSeatsFn := func() {
+		defer qs.removeQueueIfEmptyLocked(r)
+
+		qs.totSeatsInUse -= r.Seats()
+		metrics.AddRequestConcurrencyInUse(qs.qCfg.Name, r.fsName, -r.Seats())
+		if r.queue != nil {
+			r.queue.seatsInUse -= r.Seats()
+		}
+	}
+
+	defer func() {
+		if r.workEstimate.AdditionalLatency <= 0 {
+			// release the seats allocated to this request immediately
+			releaseSeatsFn()
+			return
+		}
+
+		// spin up a goroutine that sleeps for AdditionalLatency and then releases
+		// the seats allocated to this request, this ensures that the additional
+		// latency has no impact on the user experience.
+		additionalLatency := r.workEstimate.AdditionalLatency
+		go func() {
+			qs.clock.Sleep(additionalLatency)
```

Comment:
If possible I would like to avoid this goroutine but I don't see a way for it. Also - additionalLatency should be 0 for huge majority of the calls, so it's probably not that big issue.


That said - I would like to get @MikeSpreitzer comment if this goroutine will not break the goroutine counting that we currently have.

Reply:
Indeed, the queueset code should participate in goroutine counting.  A relatively small fix would be to use the EventAfterDuration method of an EventClock, that way you get efficient code in the non-testing case and goroutine counting in test code.

It should be possible to eliminate the creation of a goroutine in the case where there are no requests queued.  In this case the release of seats can be delayed until something else prods the queueSet, at which time it can be decided whether the seats should have already been released or released in the future (again either via a goroutine or waiting for another prod).

Reply:
I don't think we have many requests with `AdditionalLatency` today, so the cost of doing it in a new goroutine may be negligible today. We can let this solution merge today, and I can do a separate PR in the future that uses a `prod queueSet` approach, thoughts? I will leave a TODO if this is acceptable.